File: haskell-server/src/EagerWords/Server/Domain/Board.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE FlexibleContexts #-}

module EagerWords.Server.Domain.Board (
    Board(..)
  , mkBoard, mkEmptyBoard, mkBoardFromPieces, mkBoardFromPiecePoints
  , rows, cols
  , next, prev, adjacent
  , get, getPiece, getPiecePoints
  , setN, setBlackPoints, setPiecePoints
  , isEmpty, isFilled, inBounds
  , stripOfPlay, stripOfBoard
  , pointIsEmpty, pointIsNonEmpty
  , pointHasValue, pointHasRealNeighbor
  , validateCoordinate, validatePoint
  , stripIsDisconnectedInLine
  , playableEnclosingStripsOfBlankPoints
  , lineNeighbors
  , computeAllLiveStrips, playableStrips
)
where

import Data.List
import qualified Data.Map as Map
import Control.Monad.Except (MonadError(..))

import EagerWords.Common.Domain.PlayPiece (PlayPiece, PlayPiece(PlayPiece))
import qualified EagerWords.Common.Domain.PlayPiece as PlayPiece
import EagerWords.Common.Domain.PiecePoint (PiecePoint, PiecePoint(PiecePoint))
import qualified EagerWords.Common.Domain.PiecePoint as PiecePoint
import EagerWords.Common.Domain.Piece (Piece)
import qualified EagerWords.Common.Domain.Piece as Piece
import Bolour.Plane.Domain.Axis (Coordinate, Axis(..))
import Bolour.Plane.Domain.Point (Point, Point(Point))
import qualified Bolour.Plane.Domain.Point as Point
import Bolour.Plane.Domain.LineSegment (LineSegment, LineSegment(LineSegment))
import qualified Bolour.Plane.Domain.LineSegment as LineSegment
import qualified Bolour.Plane.Domain.Axis as Axis
import Bolour.Util.BlackWhite (BlackWhite, BlackWhite(Black, White))
import qualified Bolour.Util.BlackWhite as BlackWhite
import Bolour.Plane.Domain.BlackWhitePoint (BlackWhitePoint, BlackWhitePoint(BlackWhitePoint))
import qualified Bolour.Plane.Domain.BlackWhitePoint as BlackWhitePoint
import Bolour.Plane.Domain.BlackWhiteGrid (BlackWhiteGrid)
import qualified Bolour.Plane.Domain.BlackWhiteGrid as Gr
import EagerWords.Server.Domain.GameError (GameError(..))
import qualified Bolour.Util.MiscUtil as Util
import EagerWords.Server.Domain.Strip (Strip, Strip(Strip))
import qualified EagerWords.Server.Domain.Strip as Strip
import qualified Bolour.Util.Empty as Empty

-- | The game board.
data Board = Board {
    dimension :: Int
  , grid :: BlackWhiteGrid Piece
}
  deriving (Show)

-- TODO. Check rectangular. Check parameters. See below.
mkBoardFromPieces :: [[Maybe Piece]] -> Int -> Board
mkBoardFromPieces cells =
  let cellMaker row col = White $ cells !! row !! col
  in mkBoard cellMaker

mkBoardFromPiecePoints :: [PiecePoint] -> Int -> Board
mkBoardFromPiecePoints piecePoints dimension =
  let maybePiecePoint row col = find (\PiecePoint {point} -> point == Point row col) piecePoints
      pieceMaker row col =
        case maybePiecePoint row col of
          Nothing -> White Nothing
          Just (PiecePoint piece point) -> White (Just piece)
  in mkBoard pieceMaker dimension

mkEmptyBoard :: Int -> Board
mkEmptyBoard dimension =
  let grid = Gr.mkEmptyGrid dimension dimension
  in Board dimension grid

mkBoard :: (Int -> Int -> BlackWhite Piece) -> Int -> Board
mkBoard cellMaker dimension =
  let grid = Gr.mkGrid cellMaker dimension dimension
  in Board dimension grid

rows :: Board -> [[BlackWhitePoint Piece]]
rows Board {grid} = Gr.rows grid

cols :: Board -> [[BlackWhitePoint Piece]]
cols Board {grid} = Gr.cols grid

isFilled :: Board -> Bool
isFilled Board {grid} = Gr.isFilled grid

boardLine :: Board -> Axis -> Int -> [BlackWhitePoint Piece]
boardLine board Axis.X lineNumber = rows board !! lineNumber
boardLine board Axis.Y lineNumber = cols board !! lineNumber

blackWhitePointsToStrip :: Axis -> Int -> [BlackWhitePoint Piece] -> Int -> Int -> Strip
blackWhitePointsToStrip axis lineNumber blackWhitePiecePoints offset size =
  let blackWhiteChars = bwPiecePointsToBWChars blackWhitePiecePoints
  in Strip.stripFromBlackWhiteLine axis lineNumber blackWhiteChars offset size

bwPiecePointToBWChar :: BlackWhitePoint Piece -> BlackWhite Char
bwPiecePointToBWChar blackWhitePiecePoint =
  Piece.value <$> BlackWhitePoint.value blackWhitePiecePoint

bwPiecePointsToBWChars :: [BlackWhitePoint Piece] -> [BlackWhite Char]
bwPiecePointsToBWChars bwPiecePoints = bwPiecePointToBWChar <$> bwPiecePoints

stripOfBoard :: Board -> Axis -> Int -> Int -> Int -> Strip
stripOfBoard board axis lineNumber offset size =
  let theLine = boardLine board axis lineNumber
  in blackWhitePointsToStrip axis lineNumber theLine offset size

next :: Board -> Point -> Axis -> Maybe (BlackWhite Piece)
next Board {grid} point axis = BlackWhitePoint.value <$> Gr.next grid point axis

prev :: Board -> Point -> Axis -> Maybe (BlackWhite Piece)
prev Board {grid} point axis = BlackWhitePoint.value <$> Gr.prev grid point axis

adjacent :: Board -> Point -> Axis -> Int -> Maybe (BlackWhite Piece)
adjacent Board {grid} point axis direction = BlackWhitePoint.value <$> Gr.adjacent grid point axis direction

-- | Nothing if out of bounds, noPiece if empty but in bounds.
get :: Board -> Point -> BlackWhite Piece
get board @ Board { grid }  = Gr.get grid

getPiece :: Board -> Point -> Maybe Piece
getPiece board point = BlackWhite.fromWhite $ get board point

getPiecePoints :: Board -> [PiecePoint]
getPiecePoints Board {grid} =
   let locatedPieces = Gr.getValues grid
   in uncurry PiecePoint <$> locatedPieces

setN :: Board -> [BlackWhitePoint Piece] -> Board
setN Board {dimension, grid} bwPiecePoints =
  let updatedGrid = Gr.setN grid bwPiecePoints
  in Board dimension updatedGrid

setPiecePoints :: Board -> [PiecePoint] -> Board
setPiecePoints board piecePoints =
  let toBWPoint PiecePoint {piece, point} = BlackWhitePoint (White (Just piece)) point
      bwPiecePoints = toBWPoint <$> piecePoints
  in setN board bwPiecePoints

setBlackPoints :: Board -> [Point] -> Board
setBlackPoints board points =
  let blackBWPoints = BlackWhitePoint Black <$> points
  in setN board blackBWPoints

isEmpty :: Board -> Bool
isEmpty Board { grid } = Empty.isEmpty grid

pointIsEmpty :: Board -> Point -> Bool
pointIsEmpty Board {grid} = Gr.isEmpty grid

pointIsNonEmpty :: Board -> Point -> Bool
pointIsNonEmpty board point = not $ pointIsEmpty board point

pointHasValue :: Board -> Point -> Bool
pointHasValue Board {grid} = Gr.hasValue grid

inBounds :: Board -> Point -> Bool
inBounds Board {grid} = Gr.inBounds grid

validateCoordinate :: MonadError GameError m =>
  Board -> Axis -> Coordinate -> m Coordinate

validateCoordinate (board @ Board { dimension }) axis coordinate =
  if coordinate >= 0 && coordinate < dimension then return coordinate
  else throwError $ PositionOutOfBoundsError axis (0, dimension) coordinate

validatePoint :: MonadError GameError m =>
  Board -> Point -> m Point

validatePoint board (point @ Point { row, col }) = do
  _ <- validateCoordinate board Y row
  _ <- validateCoordinate board X col
  return point

maybeBlackWhiteHasPiece :: Maybe (BlackWhite Piece) -> Bool
maybeBlackWhiteHasPiece Nothing = False
maybeBlackWhiteHasPiece (Just bwPiece) = BlackWhite.hasValue bwPiece

pointHasRealNeighbor :: Board -> Point -> Axis -> Bool
pointHasRealNeighbor board point axis =
  let maybeNext = next board point axis
      maybePrev = prev board point axis
  in maybeBlackWhiteHasPiece maybeNext || maybeBlackWhiteHasPiece maybePrev

stripOfPlay :: Board -> [PlayPiece] -> Maybe Strip
stripOfPlay board [] = Nothing
stripOfPlay board [playPiece] = stripOfPlay1 board playPiece
stripOfPlay board (pp:pps) = stripOfPlayN board (pp:pps)

stripOfPlay1 :: Board -> PlayPiece -> Maybe Strip
stripOfPlay1 board PlayPiece {point} =
  let Point {row, col} = point
      -- Arbitrarily choose the row of the single move play.
      theRow = rows board !! row
      rowAsBlackWhiteChars = bwPiecePointsToBWChars theRow
  in Just $ Strip.stripFromBlackWhiteLine Axis.X row rowAsBlackWhiteChars col 1

stripOfPlayN :: Board -> [PlayPiece] -> Maybe Strip
stripOfPlayN board playPieces =
  let points = (\PlayPiece {point} -> point) <$> playPieces
      Point {row = hdRow, col = hdCol} = head points
      maybeAxis = Point.axisOfLine points
  in
    let mkStrip axis =
          let (lineNumber, line, begin) =
                case axis of
                  Axis.X -> (hdRow, rows board !! hdRow, hdCol)
                  Axis.Y -> (hdCol, cols board !! hdCol, hdRow)
              lineAsBlackWhiteChars = bwPiecePointsToBWChars line
           in Strip.stripFromBlackWhiteLine axis lineNumber lineAsBlackWhiteChars begin (length points)
    in mkStrip <$> maybeAxis

-- | Check that a strip has no neighbors on either side - is disconnected
--   from the rest of its line. If it is has neighbors, it is not playable
--   since a matching word will run into the neighbors. However, a containing
--   strip will be playable and so we can forget about this strip.
stripIsDisconnectedInLine :: Board -> Strip -> Bool
stripIsDisconnectedInLine board (strip @ Strip {axis, begin, end, content})
  | null content = False
  | otherwise =
      let f = Strip.stripPoint strip 0
          l = Strip.stripPoint strip (end - begin)
          -- limit = dimension
          maybePrevPiece = prev board f axis
          maybeNextPiece = next board l axis
          isSeparator maybeBlackWhitePiece = not $ maybeBlackWhiteHasPiece maybeBlackWhitePiece
      in isSeparator maybePrevPiece && isSeparator maybeNextPiece

playableStrips :: Board -> Int -> [Strip]
playableStrips board trayCapacity =
  if isEmpty board
    then playableStripsForEmptyBoard board trayCapacity
    else playableStripsForNonEmptyBoard board trayCapacity

playableStripsForEmptyBoard :: Board -> Int -> [Strip]
playableStripsForEmptyBoard board @ Board {dimension} trayCapacity =
  let center = dimension `div` 2
      centerRow = rows board !! center
      centerRowAsBlackWhiteChars = bwPiecePointsToBWChars centerRow
      strips = Strip.allStripsInBlackWhiteLine Axis.X center centerRowAsBlackWhiteChars
      includesCenter Strip {begin, end} = begin <= center && end >= center
      withinTraySize Strip {begin, end} = end - begin < trayCapacity
  in (filter includesCenter . filter withinTraySize) strips

playableStripsForNonEmptyBoard :: Board -> Int -> [Strip]
playableStripsForNonEmptyBoard board trayCapacity =
  let strips = computeAllLiveStrips board
      playableBlanks Strip {blanks} = blanks > 0 && blanks <= trayCapacity
      playables = filter playableBlanks strips
      playables' = filter Strip.hasAnchor playables
      playables'' = filter (stripIsDisconnectedInLine board) playables'
   in playables''

computeAllLiveStrips :: Board -> [Strip]
computeAllLiveStrips Board {grid} = lineSegmentToStrip <$> Gr.allSegments grid

lineSegmentToStrip :: LineSegment Piece -> Strip
lineSegmentToStrip lineSegment @ LineSegment {axis, lineNumber, begin, end, segment} =
  let charLineSegment = Piece.value <$> lineSegment
  in Strip.charLineSegmentToStrip charLineSegment

enclosingStripsOfBlankPoints :: Board -> Axis -> Map.Map Point [Strip]
enclosingStripsOfBlankPoints Board {grid} axis =
  let liveStrips = lineSegmentToStrip <$> Gr.segmentsAlongAxis grid axis
      stripsEnclosingBlanks = filter Strip.hasBlanks liveStrips
  in Util.inverseMultiValuedMapping Strip.blankPoints stripsEnclosingBlanks

playableEnclosingStripsOfBlankPoints :: Board -> Axis -> Map.Map Point [Strip]
playableEnclosingStripsOfBlankPoints board axis =
  let enclosing = enclosingStripsOfBlankPoints board axis
      playable strip @ Strip {blanks, content} =
        -- blanks <= trayCapacity &&
          stripIsDisconnectedInLine board strip &&
          length content > 1 -- Can't play to a single blank strip - would have no anchor.
  in
    filter playable <$> enclosing

-- | Get all the colinear neighbors in a given direction along a given axis
--   ordered in increasing value of the line index (excluding the point
--   itself). A colinear point is considered a neighbor if it has a real value,
--   and is adjacent to the given point, or recursively adjacent to a neighbor.
lineNeighbors :: Board -> Point -> Axis -> Int -> [PiecePoint]
lineNeighbors board @ Board {grid} point axis direction =
  let piecePointPairs = Gr.lineNeighbors grid point axis direction
  in toPiecePoint <$> piecePointPairs
     where toPiecePoint (piece, point) = PiecePoint piece point

File: scala-server/app/com/bolour/eagerwords/server/domain/Board.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords.com/blob/master/LICENSE.md
 */
package com.bolour.eagerwords.server.domain

import com.bolour.plane.domain.Axis.Axis
import com.bolour.app.kernel.server.domain.KernelExceptions.InternalAppException
import com.bolour.util.CommonUtil.inverseMultiValuedMapping
import com.bolour.eagerwords.common.domain.{Piece, PiecePoint, PlayPiece}
import com.bolour.plane.domain._
import com.bolour.util.{Black, BlackWhite, White}

/**
  * The game board.
  */
case class Board(dimension: Int, grid: BlackWhiteGrid[Piece]) {

  import Board._

  /**
    * Is the board empty? Empty means all of its points are White(None)?
    * Note that a black point makes the board non-empty in this
    * interpretation.
    */
  def isEmpty: Boolean = grid.isEmpty

  def isFilled: Boolean = grid.isFilled

  /**
    * If there is a piece at the given point return it in Some, otherwise
    * return None.
    */
  def getPiece(point: Point): Option[Piece] = {
    val bw = grid.get(point)
    bw match {
      case White(Some(piece)) => Some(piece)
      case _ => None
    }
  }

  /**
    * Get the value of a point on the board (if the point is out of bounds return black -
    * same as if the point is black).
    */
  def get(point: Point): BlackWhite[Piece] = grid.get(point)

  /**
    * Get the board pieces and their locations as grid pieces.
    */
  def piecePoints: List[PiecePoint] = {
    val piecesAndPoints = grid.getValues
    piecesAndPoints map { case (piece, point) => PiecePoint(piece, point)}
  }

  /**
    * Update a number of values of the board returning the updated board.
    * @param bwPoints List of (piece, point) pairs (housed in BlackWhitePoints) to update.
    *                 Out of bound points are ignored.
    */
  def setN(bwPoints: List[BlackWhitePoint[Piece]]): Board = {
    val augmentedGrid = grid.setN(bwPoints)
    Board(dimension, augmentedGrid)
  }

  /**
    * Similar to setN but provides the (piece, point) pairs in a list of PiecePoint.
    */
  def setPiecePoints(piecePoints: List[PiecePoint]): Board = {
    val bwPoints = piecePoints map { case PiecePoint(piece, point) => BlackWhitePoint(White(Some(piece)), point)}
    setN(bwPoints)
  }

  /**
    * Set a number of the board's points to black.
    */
  def setBlackPoints(deadPoints: List[Point]): Board = {
    val bwPoints = deadPoints map { point => BlackWhitePoint(Black[Piece](), point) }
    setN(bwPoints)
  }

  private def rows = grid.rows
  private def columns = grid.columns

  def boardLine(axis: Axis, lineNumber: Int): List[BlackWhitePoint[Piece]] =
    axis match {
      case Axis.X => rows(lineNumber)
      case Axis.Y => columns(lineNumber)
    }

  private def bwPiecePointToBWChar(blackWhitePiecePoint: BlackWhitePoint[Piece]): BlackWhite[Char] =
    blackWhitePiecePoint.value map { _.value }

  private def bwPiecePointsToBWChars(blackWhitePiecePoints: List[BlackWhitePoint[Piece]]): List[BlackWhite[Char]] =
    blackWhitePiecePoints map bwPiecePointToBWChar

  def stripOfBoard(axis: Axis, lineNumber: Int, offset: Int, size: Int): Strip = {
    val blackWhiteChars = bwPiecePointsToBWChars(boardLine(axis, lineNumber))
    Strip.stripFromBlackWhiteLine(axis, lineNumber, blackWhiteChars, offset, size)
  }

  def deadPoints(): List[Point] = this.grid.blackPoints()

  /**
    * Given a play in terms of a list of play pieces, find the strip
    * on which the play is made.
    *
    * TODO. Assumes play is contiguous.
    */
  def stripOfPlay(playPieces: List[PlayPiece]): Strip = {
    val len = playPieces.length
    if (len == 0)
      throw new InternalAppException("no pieces in play", null)

    if (len == 1) {
      val PlayPiece(_, point, _) = playPieces.head
      // Arbitrarily consider the single play it as a horizontal play.
      val theRow = rows(point.row)
      val rowAsBlackWhiteChars = bwPiecePointsToBWChars(theRow)
      return Strip.stripFromBlackWhiteLine(Axis.X, point.row, rowAsBlackWhiteChars, point.col, 1)
    }

    val points = playPieces.map(_.point)

    val head = points.head
    val next = points(1)

    val axis = if (head.row == next.row) Axis.X else Axis.Y
    val (lineNumber, line, begin) = axis match {
      case Axis.X => (head.row, rows(head.row), head.col)
      case Axis.Y => (head.col, columns(head.col), head.row)
    }
    val lineAsBlackWhiteChars = bwPiecePointsToBWChars(line)
    Strip.stripFromBlackWhiteLine(axis, lineNumber, lineAsBlackWhiteChars, begin, points.length)
  }

  def pointIsEmpty(point: Point): Boolean = grid.get(point).isEmpty

  def inBounds(point: Point): Boolean = {
    val Point(row, col) = point
    inBounds(row) && inBounds(col)
  }

  def inBounds(coordinate: Int): Boolean = coordinate >= 0 && coordinate < dimension

  def colinearPoint(point: Point, axis: Axis, direction: Int)(steps: Int): Option[Point] = {
    val nth = point.colinearPoint(axis, direction)(steps)
    if (!inBounds(nth)) None else Some(nth)
  }

  /**
    * Considering a point to be a neighbor if it is recursively adjacent to
    * the given point in the given direction, find the farthest one in that direction.
    */
  def farthestNeighbor1(point: Point, axis: Axis, direction: Int): Point = {

    def outOfBoundsOrEmpty(oPoint: Option[Point]): Boolean =
      oPoint.isEmpty || pointIsEmpty(oPoint.get)

    def adjacent(p: Point): Option[Point] = colinearPoint(p, axis, direction)(1)

    def isBoundary(p: Point): Boolean =
      !pointIsEmpty(p) && outOfBoundsOrEmpty(adjacent(p))

    // The starting point is special because it is empty.
    if (outOfBoundsOrEmpty(adjacent(point)))
      return point

    val neighbors = (1 until dimension).toList map colinearPoint(point, axis, direction)
    val farthest = neighbors find (_.exists(isBoundary))
    farthest.get.get // A boundary always exists.
  }

  def lineNeighbors(point: Point, axis: Axis, direction: Int): List[PiecePoint] = {
    val piecePointPairs = grid.lineNeighbors(point, axis, direction)
    piecePointPairs map { case (piece, pt) => PiecePoint(piece, pt) }
  }

  /**
    * Get the value of the next point on the grid (None if next is off the grid).
    */
  def next(point: Point, axis: Axis): Option[BlackWhite[Piece]] =
    grid.next(point, axis) map { _.value }

  /**
    * Get the value of the previous point on the grid (None if previous is off the grid).
    */
  def prev(point: Point, axis: Axis): Option[BlackWhite[Piece]] =
    grid.prev(point, axis) map { _.value }

  /**
    * Get the value of an adjacent point on the grid (None if adjacent is off the grid).
    */
  def adjacent(point: Point, axis: Axis, direction: Int): Option[BlackWhite[Piece]] =
    grid.adjacent(point, axis, direction) map { _.value }


  def hasRealNeighbor(point: Point, axis: Axis): Boolean = {
    val nextOpt = next(point, axis)
    if (nextOpt.isDefined && nextOpt.get.hasValue)
      return true
    val prevOpt = prev(point, axis)
    if (prevOpt.isDefined && prevOpt.get.hasValue)
      return true
    false
  }

  def playableEmptyStrips(traySize: Int): List[Strip] = {
    val center = dimension/2
    val strips = grid.segmentsInLine(Axis.X, center) map lineSegmentToStrip
    val conformantStrips = strips.filter { strip => strip.begin <= center && strip.end >= center}
    conformantStrips
  }

  def playableStrips(traySize: Int): List[Strip] = {
    val allStrips = computeAllLiveStrips
    def hasFillableBlanks = (s: Strip) => s.numBlanks > 0 && s.numBlanks <= traySize
    val conformantStrips1 = allStrips.filter(hasFillableBlanks)
    val conformantStrips2 = conformantStrips1.filter(_.hasAnchor)
    val conformantStrips3 = conformantStrips2.filter(stripIsDisconnectedInLine)
    conformantStrips3
  }

  def stripIsDisconnectedInLine(strip: Strip): Boolean = {
    val firstPoint = strip.point(0)
    val lastPoint = strip.point(strip.end - strip.begin)
    val maybePrevPiece = prev(firstPoint, strip.axis)
    val maybeNextPiece = next(lastPoint, strip.axis)

    def isSeparator(maybeBlackWhitePiece: Option[BlackWhite[Piece]]): Boolean = {
      maybeBlackWhitePiece match {
        case None => true
        case Some(bwPiece) =>
          bwPiece match {
            case Black() => true
            case White(None) => true
            case White(Some(_)) => false
          }
      }
    }
    isSeparator(maybePrevPiece) && isSeparator(maybeNextPiece)
  }

  def computeAllLiveStrips : List[Strip] = grid.allSegments map lineSegmentToStrip

  def enclosingStripsOfBlankPoints(axis: Axis): Map[Point, List[Strip]] = {
    val stripsEnclosingBlanks = (grid.segmentsAlongAxis(axis) map lineSegmentToStrip) filter { _.numBlanks > 0 }
    inverseMultiValuedMapping((strip: Strip) => strip.blankPoints)(stripsEnclosingBlanks)
  }

  def playableEnclosingStripsOfBlankPoints(axis: Axis): Map[Point, List[Strip]] = {
    val enclosing = enclosingStripsOfBlankPoints(axis)
    val playable = enclosing mapValues { (strips: List[Strip]) =>
      strips filter { (s: Strip) =>
          stripIsDisconnectedInLine(s) &&
          // Can't play to a single blank strip - would have no anchor.
          s.content.length > 1
      }
    }
    playable
  }
}

object Board {
  def apply(dimension: Int, cellMaker: Int => Int => BlackWhite[Piece]) : Board = {
    val grid = BlackWhiteGrid[Piece](cellMaker, dimension, dimension)
    Board(dimension, grid)
  }

  def apply(dimension: Int) : Board = {
    def cellMaker(row: Int)(col: Int): BlackWhite[Piece] = White(None)
    val grid = BlackWhiteGrid[Piece](cellMaker _, dimension, dimension)
    Board(dimension, grid)
  }

  /**
    * Create a board with the given pieces at the given positions in the grid pieces.
    */
  def apply(dimension: Int, piecePoints: List[PiecePoint]): Board = {
    def maybePiecePoint(r: Int, c: Int) = piecePoints.find(_.point == Point(r, c))
    def cellMaker(row: Int)(col: Int): BlackWhite[Piece] = {
      maybePiecePoint(row, col) match {
        case Some(PiecePoint(piece, point)) => White(Some(piece))
        case None => White(None)
      }
    }
    Board(dimension, cellMaker _)
  }

  def lineSegmentToStrip(lineSegment: LineSegment[Piece]): Strip = {
    def optToChar(opt: Option[Piece]): Char =
      opt match {
        case None => ' '
        case Some(piece) => piece.value
      }

    lineSegment match {
      case LineSegment(axis, lineNumber, begin, end, segment) =>
        Strip(axis, lineNumber, begin, end, (segment map optToChar).mkString)
    }
  }
}

